### ISSUE TITLE
Regression related to splice and something else (Spit module)

### ISSUE THREAD

**Issue by AlexDaniel** [rakudo/rakudo#1843]
No golf yet, so currently requires running tests in [Spit](https://github.com/spitsh/spitsh) module.

Here's the timeline:
* It worked in 2018.04.1 and up to `a5d46b4273b^`
* It stopped working after a5d46b4273b.
* This continued up to 3642a879d99, in which a5d46b4273b was reverted. However, the module is still failing, but with a slightly different error message.

Output on all relevant commits: https://gist.github.com/b3e8087c646cb6698ab5cf5d13f49a59

Basically, commit a5d46b4273b shadows **some other commit that causes the current problem**. It can be bisected by reverting a5d46b4273b on every step.

**Comment by AlexDaniel** [comment]
BTW every test file fails, I was running https://github.com/spitsh/spitsh/blob/master/t/compiling.t for my tests.

**Comment by jstuder-gh** [comment]
Looks like nqp::slice is causing problems :(

@AlexDaniel, @niner, I was able to track down the other commit (that a5d46b4273bc5cb011f8f9573dd4e553c5256302 was obscuring). I rebased master and removed the a5d46b4273bc5cb011f8f9573dd4e553c5256302 commit ([branch can be seen here](https://github.com/jstuder-gh/rakudo/tree/master_rebase_spit_bisect_05_2018)) and bisected with [this script](https://gist.github.com/jstuder-gh/9949e882a6d543f1f9f44d8367699fda). It states that the following commit is the first failure:

    fca7b0026373237de089ae582468def25ff3d991 is the first bad commit
    commit fca7b0026373237de089ae582468def25ff3d991
    Author: Stefan Seifert <nine@detonation.org>
    Date:   Fri May 18 16:04:59 2018 +0200

        Fix initialization of our scoped arrays
    
        There's no need to create a new Array instance and bind it. We already
        have an object that we just have to fill with data. Binding a new object  
        worked only for the local variable, but didn't affect the Stash entry.
    
    :040000 040000 35f7a0632a2fa382a14fbab7b5cff3417dae5059 cad6a9468f6a403699e0324d5330725d8a476be1 M  src
    bisect run success

(Note that the commit hash is different due to this being the rebased branch).

So by reverting fb0654cfb3bca6b8bcbf9ed01fc757871d5876df, 54e390f57e5e05bb27af7e9c19119c4a2cc6fffa, and 073a7f50f9dcd17e13a2d74c1c29aba4ed6717ce, (54e390f57e5e05bb27af7e9c19119c4a2cc6fffa needed to be revert because otherwise 073a7f50f9dcd17e13a2d74c1c29aba4ed6717ce didn't revert cleanly) I was able to get the Spit tests working again.

I'll try to figure out why slice and Spit don't like each other, but in the meantime hopefully this means the release can happen.

**Comment by jstuder-gh** [comment]
I'm not sure if they should be reverted either or if the release should be postponed a bit. But it is an option, I suppose.

**Comment by jstuder-gh** [comment]
I've found the issue between slice and Spit. At some point, it sends a $elems amount greater than the number of elems in the VMArray and slice's response is to die. If we limit the value sent to slice this addresses the issue. @lizmat does [this commit](https://github.com/rakudo/rakudo/compare/master...jstuder-gh:splice_slice_end) look good to you? If so I'll submit it as a PR.

**Comment by zoffixznet** [comment]
@jstuder-gh are we able to just re-use `removed` for `$end` calculation?

`nqp::slice($reified,$offset, nqp::sub_i(nqp::add_i($offset,removed),1))`

If yes, submit a PR and I'll merge if not, submit your original commit.

**Comment by jstuder-gh** [comment]
Yeah you're rigtht, we can reuse that. I've amended the commit and submitted the PR [here](https://github.com/rakudo/rakudo/pull/1845)

**Comment by jstuder-gh** [comment]
@zoffixznet, thanks for merging my commit!

I think it might be a bit too early to close this issue, though. Unfortunately the bug still exists (though we squashed one part of it!). The part we squashed was covering up another part. See my first post in this thread for more details.

I just realized that my comment in that first post was misleading. When it says that I got the Spit tests working by reverting the commits I meant in my local Rakudo instance although I didn't clarify. I didn't submit a PR / change anything as I'm not sure how you guys want to handle it (revert or fix and possibly delay release).

**Comment by AlexDaniel** [comment]
@jstuder-gh generally, there's no rush, especially when it comes to resolving blockers. So IMO fixing it properly is a better way to go, unless there are any concerns that would justify a full revert. That, however, doesn't mean it should take forever, so what's the estimated time of arrival for your PR? Or can you provide more info?

**Comment by zoffixznet** [comment]
Don't revert anything. The fix should be simple enough once the golf exists.

**Comment by zoffixznet** [comment]
Don't got the golf yet, but going to bed right now...

So far, I traced the bug to being caused by the new `optimize_array_variable_initialization` (of https://github.com/rakudo/rakudo/issues/1832 fame) in `src/Perl6/Optimizer.nqp`

And the only place it seems to be triggering is at https://github.com/spitsh/spitsh/blob/9bfc183c6c2aa01b7f30c5cbfc9b7dabd39429e8/lib/Spit/SAST.pm6#L1247-L1248 and if I add `dd [@types, $++];` before that `if @types`, the bug goes away, so seems there's something busted with reification when `ZZ1` QAST below gets rewritten to `ZZ1.1` QAST:

```
ZZ1:
- QAST::Op(callmethod STORE) <sunk nosink>
  - QAST::Var(lexical @types) <wanted sunk sinkok> @types
  - QAST::Op(callstatic &infix:<,>) <wanted> ,
    - QAST::Op(if) <wanted> if $!LHS-junct-ctx !== NEVER-RETURN
      - QAST::Op(callstatic &prefix:<!>) 
        - QAST::Op(callstatic &infix:<==>) 
          - QAST::Var(attribute $!LHS-junct-ctx) <wanted> $!LHS-junct-ctx
            - QAST::Var(lexical self) 
            - QAST::WVal(SAST::Junction) 
          - QAST::WVal(JunctionContext) <wanted> NEVER-RETURN
      - QAST::Op(hllize) <wanted>
        - QAST::Op(callmethod type)  type
          - QAST::Op(callstatic &postcircumfix:<[ ]>) <wanted> [0]
            - QAST::Var(lexical self) <wanted> self
            - QAST::Want <wanted> 0
              - QAST::WVal(Int) 
              - Ii
              - QAST::IVal(0)  0
      - QAST::WVal(Slip) <wanted>
    - QAST::Op(if) <wanted> if $!RHS-junct-ctx !== NEVER-RETURN
      - QAST::Op(callstatic &prefix:<!>) 
        - QAST::Op(callstatic &infix:<==>) 
          - QAST::Var(attribute $!RHS-junct-ctx) <wanted> $!RHS-junct-ctx
            - QAST::Var(lexical self) 
            - QAST::WVal(SAST::Junction) 
          - QAST::WVal(JunctionContext) <wanted> NEVER-RETURN
      - QAST::Op(hllize) <wanted>
        - QAST::Op(callmethod type)  type
          - QAST::Op(callstatic &postcircumfix:<[ ]>) <wanted> [1]
            - QAST::Var(lexical self) <wanted> self
            - QAST::Want <wanted> 1
              - QAST::WVal(Int) 
              - Ii
              - QAST::IVal(1)  1
      - QAST::WVal(Slip) <wanted>
  - QAST::WVal+{QAST::SpecialArg}(Bool :named<initialize>) 
```

```
ZZ1.1:
- QAST::Stmts 
  - QAST::Op(callmethod reify-until-lazy) <sunk nosink>
    - QAST::Op(getattr) <wanted> ,
      - QAST::Op(p6bindattrinvres) 
        - QAST::Op(p6bindattrinvres) 
          - QAST::Var(lexical @types) <wanted sunk sinkok> @types
          - QAST::WVal(List) 
          - QAST::SVal($!reified) 
          - QAST::Op(create) 
            - QAST::WVal(IterationBuffer) 
        - QAST::WVal(List) 
        - QAST::SVal($!todo) 
        - QAST::Op(p6bindattrinvres) 
          - QAST::Op(p6bindattrinvres) 
            - QAST::Op(p6bindattrinvres) 
              - QAST::WVal(List::Reifier) 
              - QAST::WVal(List::Reifier) 
              - QAST::SVal($!reified) 
              - QAST::Op(getattr) 
                - QAST::Var(lexical @types) <wanted sunk sinkok> @types
                - QAST::WVal(List) 
                - QAST::SVal($!reified) 
            - QAST::WVal(List::Reifier) 
            - QAST::SVal($!reification-target) 
            - QAST::Op(callmethod reification-target) 
              - QAST::Var(lexical @types) <wanted sunk sinkok> @types
          - QAST::WVal(List::Reifier) 
          - QAST::SVal($!future) 
          - QAST::Op(list) 
            - QAST::Op(if) <wanted> if $!LHS-junct-ctx !== NEVER-RETURN
              - QAST::Op(callstatic &prefix:<!>) 
                - QAST::Op(callstatic &infix:<==>) 
                  - QAST::Var(attribute $!LHS-junct-ctx) <wanted> $!LHS-junct-ctx
                    - QAST::Var(lexical self) 
                    - QAST::WVal(SAST::Junction) 
                  - QAST::WVal(JunctionContext) <wanted> NEVER-RETURN
              - QAST::Op(hllize) <wanted>
                - QAST::Op(callmethod type)  type
                  - QAST::Op(callstatic &postcircumfix:<[ ]>) <wanted> [0]
                    - QAST::Var(lexical self) <wanted> self
                    - QAST::Want <wanted> 0
                      - QAST::WVal(Int) 
                      - Ii
                      - QAST::IVal(0)  0
              - QAST::WVal(Slip) <wanted>
            - QAST::Op(if) <wanted> if $!RHS-junct-ctx !== NEVER-RETURN
              - QAST::Op(callstatic &prefix:<!>) 
                - QAST::Op(callstatic &infix:<==>) 
                  - QAST::Var(attribute $!RHS-junct-ctx) <wanted> $!RHS-junct-ctx
                    - QAST::Var(lexical self) 
                    - QAST::WVal(SAST::Junction) 
                  - QAST::WVal(JunctionContext) <wanted> NEVER-RETURN
              - QAST::Op(hllize) <wanted>
                - QAST::Op(callmethod type)  type
                  - QAST::Op(callstatic &postcircumfix:<[ ]>) <wanted> [1]
                    - QAST::Var(lexical self) <wanted> self
                    - QAST::Want <wanted> 1
                      - QAST::WVal(Int) 
                      - Ii
                      - QAST::IVal(1)  1
              - QAST::WVal(Slip) <wanted>
      - QAST::WVal(List) 
      - QAST::SVal($!todo) 
  - QAST::Var(lexical @types) <wanted sunk sinkok> @types
```

**Comment by niner** [comment]
Seems like the generated QAST is not reentrant. Golf:
> perl6 -e 'sub init() { my @a = 1, (init() unless $++); say @a }; init()'
[1]
[(Mu) (Mu) (Mu)]
